source/extensions/common/async_files/async_file_handle.h

```diff
@@ -0,0 +1,81 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "source/common/buffer/buffer_impl.h"
+
+#include "absl/status/statusor.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace AsyncFiles {
+
+class AsyncFileAction;
+
+// The context for AsyncFile operations that operate on an open file.
+// Instantiated from an AsyncFileManager.
+class AsyncFileContext : public std::enable_shared_from_this<AsyncFileContext> {
+public:
+  // Action to hard link the file that is currently open. Typically for use in tandem with
+  // createAnonymousFile to turn that file into a named file after finishing writing its contents.
+  //
+  // If cancelled before the callback is called but after creating the file, unlinks the file.
+  virtual std::function<void()> createHardLink(absl::string_view filename,
+                                               std::function<void(absl::Status)> on_complete) PURE;
+
+  // Enqueues an action to close the currently open file.
+  // It is an error to use an AsyncFileContext after calling close.
```

Comment:
Perhaps have all functions return absl::StatusOr so you can signify errors like this to the caller and have more sanity checking?

Reply:
Comment was outdated, duplicate is asynchronous now for consistency.

Reply:
I still think you should switch to StatusOr so that you can clearly indicate other precondition errors like the file is closed, etc.

Reply:
Oops, long email reply went into the whole-scope comments rather than this thread.